File: tests/spatial_conditional_probe.R
source("R/funs.R")

survey <- Sys.getenv("ATB_SURVEY", unset = "202207")
out_path <- Sys.getenv(
  "ATB_SPATIAL_CONDITIONAL_OUT",
  unset = file.path("docs", "artifacts", "spatial_conditional_r.csv")
)

set.seed(as.integer(survey))

surveydir <- file.path("surveydata", survey)
sfiles <- read_survey_files(surveydir)
scaling_classes <- c("SS1", "SS2", "BT")

surveydata <- ATSurveyData(
  acoustics = subset(sfiles$acoustics, class %in% scaling_classes),
  scaling = sfiles$scaling,
  age_length = sfiles$age_length,
  length_weight = sfiles$length_weight,
  trawl_locations = sfiles$trawl_locations,
  grid = sfiles$surveygrid,
  dA = (10 * km2nmi)^2
)

atbp <- ATBootstrapProblem(surveydata, scaling_classes = scaling_classes)

.head_string <- function(x, n = 8) {
  x <- as.integer(x)
  if (length(x) == 0) {
    return("")
  }
  paste(utils::head(x, n), collapse = " ")
}

.summary_stats <- function(x, prefix) {
  x <- as.numeric(x)
  tibble(
    !!paste0(prefix, "_n") := length(x),
    !!paste0(prefix, "_mean") := mean(x),
    !!paste0(prefix, "_sd") := stats::sd(x),
    !!paste0(prefix, "_min") := min(x),
    !!paste0(prefix, "_max") := max(x),
    !!paste0(prefix, "_nonpos_n") := sum(x <= 0)
  )
}

.derive_mu_z <- function(params, eps = .Machine$double.eps^(1 / 3)) {
  shift <- pmax(params$mu_x, eps)
  target_mean <- mean(params$data, na.rm = TRUE)
  if (!is.finite(target_mean) || target_mean <= 0) {
    target_mean <- mean(shift, na.rm = TRUE)
  }
  scaled_shift <- shift * target_mean / mean(shift, na.rm = TRUE)
  mu_z <- as.vector(forwardsolve(params$L, scaled_shift, upper.tri = FALSE, transpose = FALSE))
  mu_z[mu_z <= eps] <- eps
  list(scaled_shift = scaled_shift, mu_z = mu_z)
}

.nearest_domain_index <- function(domain_xy, obs_xy) {
  d <- .cross_distance_matrix(obs_xy, domain_xy)
  max.col(-d, ties.method = "first")
}

metrics <- bind_rows(lapply(atbp$class_problems, function(cp) {
  params <- cp$params
  obs_xy <- as.matrix(cp$geosetup$observed[, c("x", "y"), drop = FALSE])
  dom_xy <- as.matrix(cp$geosetup$domain[, c("x", "y"), drop = FALSE])
  nn_idx <- .nearest_domain_index(dom_xy, obs_xy)
  derived <- .derive_mu_z(params)
  ldiag <- diag(params$L)
  implied_var <- rowSums(params$L * params$L)

  bind_cols(
    tibble(
      backend = "r",
      survey = survey,
      class = cp$class,
      domain_n = nrow(dom_xy),
      obs_n = nrow(obs_xy),
      mapped_dloc_n = length(unique(nn_idx)),
      mapped_collision_n = nrow(obs_xy) - length(unique(nn_idx)),
      params_data_n = length(params$data),
      params_dloc_n = length(params$dlocs),
      params_sloc_n = length(params$slocs),
      dloc_head = .head_string(params$dlocs),
      sloc_head = .head_string(params$slocs),
      data_mean = mean(params$data),
      data_sd = stats::sd(params$data),
      mu = params$mu
    ),
    .summary_stats(params$mu_x, "prep_shift"),
    .summary_stats(derived$scaled_shift, "scaled_shift"),
    .summary_stats(derived$mu_z, "mu_z"),
    .summary_stats(ldiag, "L_diag"),
    .summary_stats(implied_var, "implied_var")
  )
}))

dir.create(dirname(out_path), recursive = TRUE, showWarnings = FALSE)
readr::write_csv(metrics, out_path)
cat("Wrote R spatial conditional metrics to ", out_path, "\n", sep = "")

File: tests/spatial_conditional_probe.jl
using CSV, DataFrames, DataFramesMeta
using LinearAlgebra
using Statistics, StatsBase
using StableRNGs

include(joinpath(@__DIR__, "..", "src", "ATBootstrap.jl"))
import .ATBootstrap as ATB

survey = get(ENV, "ATB_SURVEY", "202207")
out_path = get(
    ENV,
    "ATB_SPATIAL_CONDITIONAL_OUT",
    joinpath(@__DIR__, "..", "docs", "artifacts", "spatial_conditional_julia.csv"),
)

StableRNGs.seed!(parse(Int, survey))

surveydir = joinpath(@__DIR__, "..", "surveydata", survey)
const km2nmi = 1 / 1.852
resolution = 10.0
dA = (resolution * km2nmi)^2
scaling_classes = ["SS1", "SS2", "BT"]

(; acoustics, scaling, age_length, length_weight, trawl_locations, surveygrid) = ATB.read_survey_files(surveydir)
acoustics = @subset(acoustics, in(scaling_classes).(:class))

surveydata = ATB.ATSurveyData(
    acoustics,
    scaling,
    age_length,
    length_weight,
    trawl_locations,
    surveygrid,
    dA,
)

atbp = ATB.ATBootstrapProblem(surveydata; scaling_classes=scaling_classes)

head_string(x, n=8) = isempty(x) ? "" : join(first(collect(x), min(length(x), n)), " ")

function summary_stats(x, prefix)
    x = collect(Float64, x)
    DataFrame(
        Symbol(prefix * "_n") => [length(x)],
        Symbol(prefix * "_mean") => [mean(x)],
        Symbol(prefix * "_sd") => [std(x)],
        Symbol(prefix * "_min") => [minimum(x)],
        Symbol(prefix * "_max") => [maximum(x)],
        Symbol(prefix * "_nonpos_n") => [count(<=(0), x)],
    )
end

function derive_mu_z(params; ϵ=cbrt(eps()))
    shift = copy(params.μx)
    shift[shift .< ϵ] .= ϵ
    target_mean = mean(params.data)
    if !isfinite(target_mean) || target_mean <= 0
        target_mean = mean(shift)
    end
    scaled_shift = shift .* target_mean ./ mean(shift)
    μz = params.L \ scaled_shift
    μz[μz .<= ϵ] .= ϵ
    (; scaled_shift, μz)
end

metrics = map(atbp.class_problems) do cp
    params = cp.params
    obs_nasc = @subset(surveydata.acoustics, :class .== cp.class)
    derived = derive_mu_z(params)
    L = Matrix(params.L)
    ldiag = diag(L)
    implied_var = vec(sum(abs2, L; dims=2))

    hcat(
        DataFrame(
            backend = "julia",
            survey = survey,
            class = cp.class,
            domain_n = length(params.dlocs) + length(params.slocs),
            obs_n = nrow(obs_nasc),
            mapped_dloc_n = length(params.dlocs),
            mapped_collision_n = nrow(obs_nasc) - length(params.dlocs),
            params_data_n = length(params.data),
            params_dloc_n = length(params.dlocs),
            params_sloc_n = length(params.slocs),
            dloc_head = head_string(params.dlocs),
            sloc_head = head_string(params.slocs),
            data_mean = mean(params.data),
            data_sd = std(params.data),
            mu = params.μ,
        ),
        summary_stats(params.μx, "prep_shift"),
        summary_stats(derived.scaled_shift, "scaled_shift"),
        summary_stats(derived.μz, "mu_z"),
        summary_stats(ldiag, "L_diag"),
        summary_stats(implied_var, "implied_var"),
    )
end

mkpath(dirname(out_path))
CSV.write(out_path, vcat(metrics...))
println("Wrote Julia spatial conditional metrics to ", out_path)
